src/server/fedavg.py

```diff
@@ -495,10 +509,82 @@ def test(self):
                 if len(self.test_clients) > 0:
                     self.trainer.test(self.test_clients, results["test_clients"])
 
-            self.test_results[self.current_epoch + 1] = results
+            if self.current_epoch + 1 not in self.test_results:
+                self.test_results[self.current_epoch + 1] = results
+            else:
+                self.test_results[self.current_epoch + 1].update(results)
 
         self.testing = False
 
+    def test_server(self):
+        """The function for testing FL method's output (a single global model
+        or personalized client models)."""
+        self.testing = True
+        if self.args.common.buffers == "local":
+            metrics = self.evaluate()
+            self.test_results[self.current_epoch + 1]["server"] = {"after" : metrics}
+        else:
+            # compute global model's metrics from clients metrics
+            if self.val_clients == self.train_clients == self.test_clients:
+                pass
+        
+        metrics = self.evaluate()
```

Comment:
So if methods with local buffers will do the evaluations twice? Maybe line 531-535 have wrong indentation.

Reply:
This was an oversight from my testing, I will fix that!